cdap-metrics-sidecar/src/main/java/io/cdap/cdap/jmx/metrics/JMXMetricsCollector.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.jmx.metrics;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.AbstractScheduledService;
+import com.google.inject.Inject;
+import com.sun.management.OperatingSystemMXBean;
+import io.cdap.cdap.api.Environment;
+import io.cdap.cdap.api.metrics.MetricsCollectionService;
+import io.cdap.cdap.api.metrics.MetricsCollector;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.proto.id.NamespaceId;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.management.ManagementFactory;
+import java.lang.management.MemoryMXBean;
+import java.lang.management.MemoryUsage;
+import java.lang.management.ThreadMXBean;
+import java.net.MalformedURLException;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.management.MBeanServerConnection;
+import javax.management.remote.JMXConnector;
+import javax.management.remote.JMXConnectorFactory;
+import javax.management.remote.JMXServiceURL;
+
+/**
+ * Represents a service  that runs along with other services to collect resource usage metrics and publish them to
+ * {@link MetricsCollectionService}. For this service to work, the jvm process needs to expose JMX server on the same
+ * port that this service polls. To do this, the following JAVA OPTS need to be set:
+ * {@code -Dcom.sun.management.jmxremote=true -Dcom.sun.management.jmxremote.port=11022
+ * -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote.authenticate=false}
+ * This service requires {@code SERVICE_NAME}  env variable to be set which provides
```

Comment:
Why use environment variable instead of getting the config from the `CConfiguration`?

Reply:
This service will run alongside other services to report their resource usage. So we need a way to identify whose resource usage is being measured and tag metrics accordingly. The way I could think of doing this was using env variables. Is there a way to achieve this without env vars?

Reply:
For this class, it can be just a parameter to the constructor. From the caller (which is the service runs in k8s), it can get it by the pod name from the PodInfo?

Reply:
Changes made:

- Fetch podname from PodInfo through masterEnv and parse it to get component name.
- Pass component name as a param to constructor of JMX metrics collector. Using assisted inject to support this.